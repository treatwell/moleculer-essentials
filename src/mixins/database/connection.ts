import {
  Collection,
  CollectionOptions,
  MongoClient,
  CreateCollectionOptions,
  MongoError,
} from 'mongodb';
import { wrapMixin } from '../../types/index.js';
import { GlobalStoreMixin } from '../global-store.mixin.js';

// Declare the global variables used for MongoDB connection
declare global {
  var __MONGO_URI__: string | undefined;
  var __MONGO_DB_NAME__: string | undefined;
}

export type DatabaseConnectionOptions = {
  /**
   * Name of the database to use.
   * If not specified, will use the default one (inferred from uri).
   *
   * OVERRIDDEN by globalThis.__MONGO_DB_NAME__ if set, which is useful for tests.
   */
  databaseName?: string;
  /**
   * Name of the collection in the DB.
   */
  collectionName: string;
  /**
   * Collection creation options.
   * If not specified, will use the default one.
   * Note that it shouldn't be changed without being sure of what you are doing.
   */
  createCollectionOptions?: CreateCollectionOptions;

  /**
   * URI of the MongoDB server.
   * If not specified, will try to use one from environment:
   * - process.env.MONGO_URL
   * - process.env.MONGODB_URL
   * - 'mongodb://localhost:27017' (default)
   *
   * OVERRIDDEN by globalThis.__MONGO_URI__ if set, which is useful for tests.
   */
  uri?: string;
};

export function DatabaseConnectionMixin<
  TSchema extends Record<string, unknown> = never,
>(opts: DatabaseConnectionOptions) {
  const { databaseName, collectionName, createCollectionOptions } = opts;
  const uri =
    globalThis.__MONGO_URI__ ||
    opts.uri ||
    process.env.MONGO_URL ||
    process.env.MONGODB_URL ||
    'mongodb://localhost:27017';

  // Allow override the DB name for testing purposes.
  const dbName = globalThis.__MONGO_DB_NAME__
    ? `${globalThis.__MONGO_DB_NAME__}-${databaseName}`
    : databaseName;

  // Key is the uri, but we may want to add other options in the future
  const key = uri;

  return wrapMixin({
    mixins: [GlobalStoreMixin<MongoClient>()],

    methods: {
      getMongoClient(): MongoClient {
        return this.mongoClient;
      },
      getCollection(options?: CollectionOptions): Collection<TSchema> {
        return this.getMongoClient()
          .db(dbName)
          .collection<TSchema>(collectionName, options);
      },
    },
    created() {
      let client = this.getFromStore('mongodb', key);
      this.logger.debug(
        'MongoDB mixin starting, loading mongo client from store',
      );
      if (!client) {
        this.logger.info(
          "Didn't find mongo client in store, creating a new one",
        );
        client = new MongoClient(uri);
        this.setClientToStore('mongodb', key, client, async () => {
          this.logger.debug('Closing mongoDB connection');
          await client?.close();
          this.logger.info('MongoDB connection closed');
        });

        client.on('error', err => this.logger.error('MongoDB error', err));
      }
      this.mongoClient = client;
    },
    async started() {
      // Mongo driver already have a lock that will return the same promise if it's already connecting
      this.logger.debug('Service connecting to mongoDB');
      await this.getMongoClient().connect();
      this.logger.debug('Service connected to mongoDB, creating collection');
      try {
        await this.getMongoClient()
          .db(dbName)
          .createCollection(collectionName, createCollectionOptions);
      } catch (err) {
        // Code 48 === Collection already exists
        if ((err as MongoError)?.code !== 48) {
          this.logger.error('Error while creating collection', { err });
        }
      }
    },
    async stopped() {
      await this.removeServiceFromStore('mongodb', key);
    },
  });
}
