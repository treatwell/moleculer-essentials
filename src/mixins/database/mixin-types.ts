import type { Document, InferIdType, ObjectId, WithoutId } from 'mongodb';
import type { ActionVisibility } from 'moleculer';
import { ZodObject, ZodType } from 'zod/v4';
import type { JSONSchemaType } from '../../json-schema/index.js';
import type { ValidationSchema } from '../../validator/types.js';
import type { ActionSchemaFactory } from './actions/shared.js';

/**
 * Utility type to extract the string keys of a type.
 */
export type KeyString<T> = Extract<keyof T, string>;

export type TenantParams<
  TSchema extends Document,
  TenantField extends KeyString<TSchema> | false,
> =
  | (TenantField extends KeyString<TSchema> ? Pick<TSchema, TenantField> : null)
  | null;

export type DatabaseMethodsOptions<
  TSchema extends Document,
  TenantField extends KeyString<TSchema> | false,
> = {
  /**
   * Name of the field that will be required to be present in all documents.
   * This will be used in read queries as a filter to ensure that the user
   * can only access documents that belong to the same tenant.
   *
   * This field can only be top-level and should probably be indexed.
   */
  tenantField: TenantField;

  /**
   * Enable soft-delete for the model.
   * The remove method will only set the deletedAt field to NOW.
   * And a new `scope` field will be read to be able to read deleted documents.
   *
   * Note that if schema has a deletedAt field, you are required to set this option to true.
   */
  softDelete: TSchema extends { deletedAt?: Date } ? true : false;

  /**
   * Enable timestamps for the model.
   * Will automatically set the createdAt and updatedAt fields on write operations.
   *
   * Note that if schema has both createdAt and updatedAt field, you are required to set this option to true.
   */
  timestamps: TSchema extends { createdAt?: Date }
    ? TSchema extends { updatedAt?: Date }
      ? true
      : false
    : false;

  /**
   * Note that this function will be applied ONLY on insert operations.
   */
  idGenerator?: (doc: WithoutId<TSchema>) => InferIdType<TSchema>;

  /**
   * Prefix used for events.
   * If not specified, will disable events.
   *
   * Here is the list of events:
   * - `${eventPrefix}.created`: Sent on insertOne and insertMany
   * - `${eventPrefix}.updated`: Sent on updateOne and replaceOne
   * - `${eventPrefix}.deleted`: Sent on deleteOne
   */
  eventPrefix?: string;

  /**
   * The sQuerySchema variable is an optional parameter used for validation.
   * This is used to parse the sQuery parameter on list and count actions.
   * If omitted, sQuery will be silently discarded.
   * It uses the same validator than on actions.
   * The schema will not be published on the openAPI.
   * See addQueryOps for easy support of some mongo operators.
   */
  sQuerySchema?: ValidationSchema | ZodType;

  /**
   * Create the actions for the database mixin.
   * Read operations:
   * - find (max public)
   * - findStream (max public)
   * - getInternal (max public)
   * - get
   * - countInternal (max public)
   * - count
   * - list
   *
   * Write operations:
   * - create
   * - update
   * - remove
   *
   * If some actions are not provided here, it probably means that they are not necessary.
   * For example, there is no `findAllStream` or `updateMany` actions. This is because they are not used
   * often and is quite specific to the related service.
   */
  actions?: DatabaseActionOptions<TSchema>;
};

type DatabaseActionVisibility<T extends DatabaseActionNames> =
  T extends DatabaseActionInternalNames
    ? Exclude<ActionVisibility, 'published'>
    : ActionVisibility;

export type DatabaseActionOptions<
  TSchema extends Document & { _id?: ObjectId | string },
> = {
  [key in DatabaseActionNames]?: {
    visibility: DatabaseActionVisibility<key>;
  } & (key extends 'list'
    ? { maxPageSize?: number; defaultPageSize?: number; defaultSort?: string[] }
    : NonNullable<unknown>) &
    (key extends 'create' ? { allowClientId?: boolean } : NonNullable<unknown>);
} & {
  schema?: JSONSchemaType<TSchema> | ZodObject;
  schemaFactory?: ActionSchemaFactory;
  schemaName?: string;
};

export type DatabaseActionInternalNames =
  | 'find'
  | 'findStream'
  | 'getInternal'
  | 'countInternal';

export type DatabaseActionPublishedNames =
  | 'count'
  | 'list'
  | 'get'
  | 'create'
  | 'update'
  | 'remove';

export type DatabaseActionNames =
  | DatabaseActionInternalNames
  | DatabaseActionPublishedNames;
