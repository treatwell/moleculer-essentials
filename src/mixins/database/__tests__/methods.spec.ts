import { beforeAll, beforeEach, describe, expect, it, vi } from 'vitest';
import { Collection, FindCursor, ObjectId } from 'mongodb';
import { Readable } from 'stream';
import { createServiceBroker } from '../../../service-broker/index.js';
import { wrapService } from '../../../types/index.js';
import { AjvValidator } from '../../../validator/index.js';
import { DatabaseConnectionMixin } from '../connection.js';
import { DatabaseMethodsMixin } from '../methods.js';
import { DatabaseMethodsOptions } from '../mixin-types.js';

describe('DB Mixin V2 methods', () => {
  const broker = createServiceBroker({
    validator: new AjvValidator<'default'>(
      {
        default: {
          useDefaults: true,
          coerceTypes: true,
          allErrors: true,
          removeAdditional: false,
          discriminator: true,
        },
      },
      'default',
    ),
  });

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const opts: DatabaseMethodsOptions<any, any> = {
    softDelete: false,
    timestamps: false,
    tenantField: false,
  };

  const svc = broker.createService(
    wrapService({
      name: 'test',
      mixins: [
        DatabaseConnectionMixin({ collectionName: 'test' }),
        DatabaseMethodsMixin(opts),
      ],
    }),
  );
  const ctx = { emit: vi.fn() };

  const col = svc.getCollection() as Collection;

  beforeAll(async () => {
    await broker.start();
    return () => broker.stop();
  });

  beforeEach(async () => {
    await col.deleteMany({});
    opts.tenantField = false;
    opts.softDelete = false;
    opts.timestamps = false;
    opts.idGenerator = undefined;
  });

  describe('prepare update filter method', () => {
    it('should do nothing if timestamp is disabled', () => {
      const changes = { $set: { foo: 'bar' } };
      const expected = { $set: { foo: 'bar' } };

      expect(svc._prepareUpdateFilter(changes, 'create')).toEqual(expected);
      expect(svc._prepareUpdateFilter(changes, 'update')).toEqual(expected);
      expect(svc._prepareUpdateFilter(changes, 'replace')).toEqual(expected);
    });

    describe('create type', () => {
      it('should add timestamps', async () => {
        opts.timestamps = true;
        const changes = { $set: { foo: 'bar' } };

        const res = await col.findOneAndUpdate(
          { _id: { $exists: false } },
          svc._prepareUpdateFilter(changes, 'create'),
          { upsert: true, returnDocument: 'after' },
        );

        expect(res).toEqual({
          _id: expect.any(ObjectId),
          foo: 'bar',
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        });
        expect(res?.createdAt).toEqual(res?.updatedAt);
      });

      it('should override timestamps if provided', async () => {
        opts.timestamps = true;
        const changes = {
          $set: {
            createdAt: new Date('2000-02-03'),
            updatedAt: new Date('2000-02-03'),
            foo: 'bar',
          },
        };

        const res = await col.findOneAndUpdate(
          { _id: { $exists: false } },
          svc._prepareUpdateFilter(changes, 'create'),
          { upsert: true, returnDocument: 'after' },
        );

        expect(res).toEqual({
          _id: expect.any(ObjectId),
          foo: 'bar',
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        });

        // Allow for some time difference between mongo and test.
        const limit = Date.now() - 500;
        expect(res?.createdAt.getTime()).toBeGreaterThan(limit);
        expect(res?.updatedAt.getTime()).toBeGreaterThan(limit);
      });
    });

    describe('update type', () => {
      it('should update updatedAt but not createdAt', async () => {
        opts.timestamps = true;
        await col.insertOne({
          createdAt: new Date('2000-02-03'),
          updatedAt: new Date('2000-02-03'),
          test: 'bar',
        });

        const changes = { $set: { foo: 'bar' } };
        const res = await col.findOneAndUpdate(
          { test: 'bar' },
          svc._prepareUpdateFilter(changes, 'update'),
          { returnDocument: 'after' },
        );

        expect(res?.updatedAt.getTime()).toBeGreaterThan(Date.now() - 500);
        expect(res?.createdAt).toEqual(new Date('2000-02-03'));
      });

      it('should remove createdAt if provided', async () => {
        opts.timestamps = true;
        await col.insertOne({
          createdAt: new Date('2000-02-03'),
          updatedAt: new Date('2000-02-03'),
          test: 'bar',
        });

        const changes = {
          $set: { foo: 'bar' },
          $inc: { createdAt: 1000 * 60 * 60 * 24 * 15 },
        };
        const res = await col.findOneAndUpdate(
          { test: 'bar' },
          svc._prepareUpdateFilter(changes, 'update'),
          { returnDocument: 'after' },
        );

        expect(res?.updatedAt.getTime()).toBeGreaterThan(Date.now() - 500);
        expect(res?.createdAt).toEqual(new Date('2000-02-03'));
      });

      it('should add createdAt if update with upsert', async () => {
        opts.timestamps = true;

        const changes = { $set: { foo: 'bar' } };
        const res = await col.findOneAndUpdate(
          { test: 'bar' },
          svc._prepareUpdateFilter(changes, 'update'),
          { returnDocument: 'after', upsert: true },
        );

        expect(res?.updatedAt.getTime()).toBeGreaterThan(Date.now() - 500);
        expect(res?.createdAt.getTime()).toBeGreaterThan(Date.now() - 500);
      });

      it.each([
        [
          [
            {
              $addFields: {
                test: 'bar',
                createdAt: new Date('2010-02-03'),
                updatedAt: new Date('2010-02-03'),
              },
            },
          ],
        ],
        [[{ $set: { createdAt: new Date('2011-02-03') } }]],
        [[{ $unset: 'createdAt' }]],
        [[{ $unset: ['createdAt'] }]],
        [[{ $unset: ['createdAt', 'test2'] }]],
        [[{ $project: { createdAt: 0 } }]],
        [[{ $project: { test: 1 } }]],
      ])('should work with pipelines %#', async changes => {
        opts.timestamps = true;
        await col.insertOne({
          createdAt: new Date('2000-02-03'),
          updatedAt: new Date('2000-02-03'),
          test: 'bar',
        });

        const res = await col.findOneAndUpdate(
          { test: 'bar' },
          svc._prepareUpdateFilter(changes, 'update'),
          { returnDocument: 'after' },
        );

        expect(res).toEqual({
          _id: expect.any(ObjectId),
          test: 'bar',
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        });
        expect(res?.updatedAt.getTime()).toBeGreaterThan(Date.now() - 500);
        expect(res?.createdAt).toEqual(new Date('2000-02-03'));
      });

      it('should prevent using replaceWith in updates', async () => {
        opts.timestamps = true;
        const changes = [{ $replaceWith: { test: 'foo' } }];

        expect.hasAssertions();
        try {
          await col.findOneAndUpdate(
            { test: 'bar' },
            svc._prepareUpdateFilter(changes, 'update'),
            { returnDocument: 'after' },
          );
        } catch (e) {
          expect(e).toMatchSnapshot();
        }
      });
    });

    describe('replace type', () => {
      it('should throw if not a pipeline', async () => {
        opts.timestamps = true;
        await col.insertOne({
          createdAt: new Date('2000-02-03'),
          updatedAt: new Date('2000-02-03'),
          test: 'bar',
        });

        const changes = { $set: { foo: 'bar' } };
        expect.hasAssertions();
        try {
          await col.findOneAndUpdate(
            { test: 'bar' },
            svc._prepareUpdateFilter(changes, 'replace'),
            { returnDocument: 'after' },
          );
        } catch (e) {
          expect(e).toMatchSnapshot();
        }
      });

      it('should keep createdAt field', async () => {
        opts.timestamps = true;
        await col.insertOne({
          createdAt: new Date('2000-02-03'),
          updatedAt: new Date('2000-02-03'),
          test: 'bar',
        });

        const changes = [{ $replaceWith: { foo: 'bar' } }];
        const res = await col.findOneAndUpdate(
          { test: 'bar' },
          svc._prepareUpdateFilter(changes, 'replace'),
          { returnDocument: 'after' },
        );

        expect(res).toEqual({
          _id: expect.any(ObjectId),
          foo: 'bar',
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        });
        expect(res?.updatedAt.getTime()).toBeGreaterThan(Date.now() - 500);
        expect(res?.createdAt).toEqual(new Date('2000-02-03'));
      });
    });
  });

  describe('tenant filter method', () => {
    it('should do nothing if no tenantField', () => {
      opts.tenantField = undefined;

      expect(svc._getTenantFilter({ foo: 'bar' })).toEqual({});
    });

    it('should throw if no params are passed', () => {
      opts.tenantField = 'foo';

      expect(() => svc._getTenantFilter(null)).toThrowErrorMatchingSnapshot();
    });

    it('should throw if params does not have field', () => {
      opts.tenantField = 'foo';

      expect(() =>
        svc._getTenantFilter({ bar: 1 }),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should not throw if strict mode is disabled and params does not have field', () => {
      opts.tenantField = 'foo';

      expect(svc._getTenantFilter({ bar: 1 }, false)).toEqual({});
    });

    it('should return the correct mongo filter', () => {
      opts.tenantField = 'foo';

      expect(svc._getTenantFilter({ foo: 'myfoo-shop' })).toEqual({
        foo: 'myfoo-shop',
      });
    });

    it('should return the correct mongo filter (not strict)', () => {
      opts.tenantField = 'foo';

      expect(svc._getTenantFilter({ foo: 'myfoo-shop' }, false)).toEqual({
        foo: 'myfoo-shop',
      });
    });
  });

  describe('soft delete method', () => {
    it('should do nothing if soft delete is disabled', () => {
      opts.softDelete = false;
      expect(svc._getSoftDeleteFilter()).toEqual({});
    });

    it('should return the correct mongo filter (scope: only-deleted)', () => {
      opts.softDelete = true;
      expect(svc._getSoftDeleteFilter('only-deleted')).toEqual({
        deletedAt: { $gte: 0 },
      });
    });

    it('should return the correct mongo filter (scope: no-deleted)', () => {
      opts.softDelete = true;
      expect(svc._getSoftDeleteFilter('no-deleted')).toEqual({
        deletedAt: null,
      });
    });

    it('should return the correct mongo filter (scope: undefined)', () => {
      opts.softDelete = true;
      expect(svc._getSoftDeleteFilter()).toEqual({ deletedAt: null });
    });

    it('should return the correct mongo filter (scope: include-deleted)', () => {
      opts.softDelete = true;
      expect(svc._getSoftDeleteFilter('include-deleted')).toEqual({});
    });

    it('should throw if unknown scope', () => {
      opts.softDelete = true;
      expect(() =>
        svc._getSoftDeleteFilter('test-deleted'),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('get query filter method', () => {
    it('should return the correct mongo filter', () => {
      opts.softDelete = true;
      opts.tenantField = 'foo';

      expect(svc._getQueryFilter({ test: 1 }, { foo: 'myfoo-shop' })).toEqual({
        test: 1,
        foo: 'myfoo-shop',
        deletedAt: null,
      });
    });

    it('should always prioritize feature filters (soft-delete, tenant)', () => {
      opts.softDelete = true;
      opts.tenantField = 'foo';

      expect(
        svc._getQueryFilter(
          { test: 1, foo: 'another-shop', deletedAt: { $gte: 50 } },
          { foo: 'myfoo-shop' },
        ),
      ).toEqual({ test: 1, foo: 'myfoo-shop', deletedAt: null });
    });

    it('should optimize query after all', () => {
      opts.tenantField = 'foo';
      expect(
        svc._getQueryFilter(
          { $or: [{ test: 1 }, { testB: 2 }] },
          { foo: 'myfoo-shop' },
        ),
      ).toEqual({
        $or: [
          { test: 1, foo: 'myfoo-shop' },
          { testB: 2, foo: 'myfoo-shop' },
        ],
      });
    });
  });

  describe('create find cursor', () => {
    it('should return a find cursor', () => {
      expect(svc._createFindCursor({}, null, {})).toBeInstanceOf(FindCursor);
    });

    it('should add query filters', async () => {
      opts.softDelete = true;
      opts.tenantField = 'foo';
      await col.insertMany([
        { foo: 'bar', test: 3, deletedAt: new Date() },
        { foo: 'bar', test: 1 },
        { foo: 'bar', test: 2 },
        { foo: 'bar', test: 3 },
        { foo: 'rab', test: 4 },
      ]);

      const cursor = svc._createFindCursor(
        { test: { $gte: 2 } },
        { foo: 'bar' },
        {},
      );

      expect(cursor).toBeInstanceOf(FindCursor);
      expect(await cursor.toArray()).toHaveLength(2);
    });

    it('should transform sort and projection options', async () => {
      await col.insertMany([
        { part: 3, test: 1 },
        { part: 1, test: 1 },
        { part: 2, test: 2 },
        { part: 4, test: 2 },
      ]);

      const cursor = svc._createFindCursor({}, null, {
        sort: ['-test', 'part'],
        fields: ['part', '-_id'],
      });

      expect(cursor).toBeInstanceOf(FindCursor);
      expect(await cursor.toArray()).toEqual([
        { part: 2 },
        { part: 4 },
        { part: 1 },
        { part: 3 },
      ]);
    });

    it('should transfer driver options', async () => {
      await col.insertMany([
        { part: 3 },
        { part: 1 },
        { part: 2 },
        { part: 4 },
      ]);

      const cursor = svc._createFindCursor({}, null, {
        skip: 2,
        sort: ['part'],
        fields: ['-_id'],
      });

      expect(cursor).toBeInstanceOf(FindCursor);
      expect(await cursor.toArray()).toEqual([{ part: 3 }, { part: 4 }]);
    });
  });

  describe('find one method', () => {
    it('should return null if no document is found', async () => {
      expect(await svc._findOne({})).toBeNull();
    });

    it('should return a single document', async () => {
      await col.insertMany([{ part: 1 }, { part: 2 }]);

      expect(await svc._findOne({}, null, { sort: ['-part'] })).toEqual({
        part: 2,
        _id: expect.any(ObjectId),
      });
    });
  });

  describe('find method', () => {
    it('should return empty array', async () => {
      expect(await svc._find({})).toEqual([]);
    });

    it('should return a all documents', async () => {
      await col.insertMany([{ part: 1 }, { part: 2 }]);

      expect(
        await svc._find({}, null, { sort: ['-part'], fields: ['-_id'] }),
      ).toEqual([{ part: 2 }, { part: 1 }]);
    });
  });

  describe('findStream method', () => {
    it('should return a single document', async () => {
      await col.insertMany([{ part: 2 }, { part: 2 }]);

      const stream = svc._findStream({}, null, { fields: ['-_id'] });

      expect.assertions(3);
      expect(stream).toBeInstanceOf(Readable);

      for await (const doc of stream) {
        expect(doc).toEqual({ part: 2 });
      }
    });
  });

  describe('countDocuments method', () => {
    it('should return the correct number of document', async () => {
      await col.insertMany([{ part: 1 }, { part: 2 }, { part: 4 }]);

      const count = await svc._countDocuments({ part: { $gte: 2 } }, null, {});

      expect(count).toEqual(2);
    });
  });

  describe('insertOne method', () => {
    it('should insert a document', async () => {
      const doc = await svc._insertOne(ctx, { foo: 'bar' });

      expect(doc).toEqual({ _id: expect.any(ObjectId), foo: 'bar' });
    });

    it('should let specify projection', async () => {
      const doc = await svc._insertOne(
        ctx,
        { foo: 'bar' },
        { fields: ['-_id'] },
      );

      expect(doc).toEqual({ foo: 'bar' });
    });

    it('should generate id if generator specified', async () => {
      let id = 0;
      // @ts-expect-error For test purpose
      opts.idGenerator = () => {
        id += 1;
        return `test-${id}`;
      };

      const docA = await svc._insertOne(ctx, { foo: 'bar' });
      const docB = await svc._insertOne(ctx, {
        _id: 'specific-id',
        foo: 'bar',
      });

      expect(docA).toEqual({ _id: 'test-1', foo: 'bar' });
      expect(docB).toEqual({ _id: 'specific-id', foo: 'bar' });
    });
  });

  describe('insertMany method', () => {
    it('should insert many documents', async () => {
      const ids = await svc._insertMany(ctx, [{ foo: 'bar' }, { foo: 'rab' }]);

      expect(ids).toEqual([expect.any(ObjectId), expect.any(ObjectId)]);
    });

    it('should return ObjectIds with forceServerObjectId', async () => {
      const ids = await svc._insertMany(ctx, [{ foo: 'bar' }, { foo: 'rab' }], {
        forceServerObjectId: true,
      });

      expect(ids).toEqual([expect.any(ObjectId), expect.any(ObjectId)]);
    });

    it('should generate id if generator specified', async () => {
      let id = 0;
      // @ts-expect-error For test purpose
      opts.idGenerator = () => {
        id += 1;
        return `test-${id}`;
      };

      const ids = await svc._insertMany(ctx, [
        { foo: 'bar' },
        { _id: 'my-id', foo: 'bar' },
        { foo: 'rab' },
      ]);
      expect(ids).toEqual(['test-1', 'my-id', 'test-2']);
    });
  });

  describe('updateOne method', () => {
    it('should return the after document', async () => {
      await col.insertMany([{ test: 1 }, { test: 2 }, { test: 3 }]);

      const doc = await svc._updateOne(ctx, { test: 2 }, null, {
        $set: { test: 4 },
      });

      expect(doc).toEqual({ test: 4, _id: expect.any(ObjectId) });
    });

    it('should return the after document (with projection)', async () => {
      await col.insertMany([{ test: 1 }, { test: 2 }, { test: 3 }]);

      const doc = await svc._updateOne(
        ctx,
        { test: 2 },
        null,
        { $inc: { test: 2 } },
        { fields: ['-_id'] },
      );

      expect(doc).toEqual({ test: 4 });
    });

    it('should allow to return before version', async () => {
      await col.insertMany([{ test: 1 }, { test: 2 }, { test: 3 }]);

      const doc = await svc._updateOne(
        ctx,
        { test: 2 },
        null,
        { $inc: { test: 2 } },
        { returnDocument: 'before', skipUpdateEvent: true },
      );

      expect(doc).toEqual({ test: 2, _id: expect.any(ObjectId) });
    });

    it('should throw if update event and before doc', async () => {
      opts.eventPrefix = 'test';
      await col.insertMany([{ test: 1 }, { test: 2 }, { test: 3 }]);

      expect.hasAssertions();
      try {
        await svc._updateOne(
          ctx,
          { test: 2 },
          null,
          { $inc: { test: 2 } },
          { returnDocument: 'before' },
        );
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('updateMany method', () => {
    it('should update many documents', async () => {
      opts.tenantField = 'foo';
      await col.insertMany([
        { test: 1, foo: 'bar' },
        { test: 2, foo: 'bar' },
        { test: 3, foo: 'bar' },
        { test: 3, foo: 'rab' },
      ]);

      const res = await svc._updateMany(
        { test: { $gte: 2 } },
        { foo: 'bar' },
        { $set: { test: 3 } },
      );

      expect(res).toEqual(
        expect.objectContaining({ matchedCount: 2, modifiedCount: 1 }),
      );
    });
  });

  describe('replaceOne method', () => {
    it('should return null if no match', async () => {
      const res = await svc._replaceOne(ctx, {}, null, {
        test: 3,
      });

      expect(res).toBeNull();
    });

    it('should replace a document', async () => {
      await col.insertOne({ test: 2, notDeclared: true });

      const res = await svc._replaceOne(
        ctx,
        { test: { $gte: 2 } },
        null,
        { test: 3 },
        { fields: ['-_id'] },
      );

      expect(res).toEqual({ test: 3 });
    });

    it('should not evaluate operator in document', async () => {
      await col.insertOne({ test: 2 });

      const res = await svc._replaceOne(ctx, { test: { $gte: 2 } }, null, {
        $inc: { test: 1 },
      });

      expect(res).toEqual({
        _id: expect.any(ObjectId),
        $inc: { test: 1 },
      });
    });

    it('should throw if update event and before doc', async () => {
      opts.eventPrefix = 'test';
      await col.insertMany([{ test: 1 }, { test: 2 }, { test: 3 }]);

      expect.hasAssertions();
      try {
        await svc._replaceOne(
          ctx,
          { test: 2 },
          null,
          { test: 7 },
          { returnDocument: 'before' },
        );
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('deleteOne method', () => {
    it('should delete document', async () => {
      await col.insertOne({ test: 2 });

      const res = await svc._deleteOne(ctx, { test: { $gte: 2 } }, null);

      expect(await col.countDocuments()).toEqual(0);
      expect(res).toEqual({ _id: expect.any(ObjectId), test: 2 });
    });

    it('should soft delete', async () => {
      opts.softDelete = true;
      await col.insertOne({ test: 2 });

      const res = await svc._deleteOne(ctx, { test: { $gte: 2 } }, null);

      expect(await col.find({}).toArray()).toEqual([
        {
          _id: expect.any(ObjectId),
          test: 2,
          deletedAt: expect.any(Date),
        },
      ]);
      expect(res).toEqual({ _id: expect.any(ObjectId), test: 2 });
    });

    it('should respect projection', async () => {
      opts.softDelete = true;
      await col.insertOne({ test: 2 });

      const res = await svc._deleteOne(ctx, { test: { $gte: 2 } }, null, {
        fields: ['-_id'],
      });

      expect(res).toEqual({ test: 2 });
    });
  });

  describe('deleteMany method', () => {
    it('should delete multiple document', async () => {
      await col.insertMany([{ test: 2 }, { test: 3 }, { test: 4 }]);

      const res = await svc._deleteMany({ test: { $gte: 3 } }, null);

      expect(await col.countDocuments()).toEqual(1);
      expect(res).toEqual(2);
    });

    it('should soft delete', async () => {
      opts.softDelete = true;
      await col.insertMany([
        { test: 2 },
        { test: 3 },
        { test: 4, deletedAt: new Date() },
      ]);

      const res = await svc._deleteMany({ test: { $gte: 3 } }, null);

      expect(res).toEqual(1);
      expect(await col.find({}, { sort: { test: 1 } }).toArray()).toEqual([
        {
          _id: expect.any(ObjectId),
          test: 2,
        },
        {
          _id: expect.any(ObjectId),
          test: 3,
          deletedAt: expect.any(Date),
        },
        {
          _id: expect.any(ObjectId),
          test: 4,
          deletedAt: expect.any(Date),
        },
      ]);
    });
  });
});
