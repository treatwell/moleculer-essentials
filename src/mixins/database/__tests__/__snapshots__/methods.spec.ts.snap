// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DB Mixin V2 methods > prepare update filter method > replace type > should throw if not a pipeline 1`] = `[Error: Replace with timestamps can only go through an aggregation pipeline with a single $replaceWith stage]`;

exports[`DB Mixin V2 methods > prepare update filter method > update type > should prevent using replaceWith in updates 1`] = `[Error: $replaceWith/$replaceRoot can't be used in an update aggregation pipeline.]`;

exports[`DB Mixin V2 methods > replaceOne method > should throw if update event and before doc 1`] = `[Error: Cannot send update event with returnDocument: before option]`;

exports[`DB Mixin V2 methods > soft delete method > should throw if unknown scope 1`] = `[Error: Unknown soft delete scope test-deleted]`;

exports[`DB Mixin V2 methods > tenant filter method > should throw if no params are passed 1`] = `[Error: Missing tenant field "foo" in params]`;

exports[`DB Mixin V2 methods > tenant filter method > should throw if params does not have field 1`] = `[Error: Missing tenant field "foo" in params]`;

exports[`DB Mixin V2 methods > updateOne method > should throw if update event and before doc 1`] = `[Error: Cannot send update event with returnDocument: before option]`;
